pyresample/geometry.py

```diff
@@ -717,37 +717,282 @@ def _do_transform(src, dst, lons, lats, alt):
+
+        # Conversion to Geocentric Cartesian (x,y,z) CRS
+        res = da.map_blocks(self._do_transform, latlong, geocent,
+                            src_lons.data,
+                            src_lats.data,
+                            da.zeros_like(src_lons),  # altitude
+                            new_axis=[2],
+                            chunks=(src_lons.chunks[0], src_lons.chunks[1], 3))
+        res = xr.DataArray(res, dims=['y', 'x', 'xyz'])
+
+        # Retrieve new centroids
+        # TODO: make it dask compatible using _upsample_centroid_dask [HELP WANTED]
+        # res1 = da.apply_along_axis(_upsample_centroid_dask,
+        #                            2,
+        #                            res.data,
+        #                            x,
+        #                            y)
+        # res1 = xr.DataArray(res1, dims=['y', 'x', 'coord'], coords=src_lons.coords)
+        res = np.stack(upsample_centroids(res[:, :, 0].data,
+                                          res[:, :, 1].data,
+                                          res[:, :, 2].data, x=x, y=y), axis=2)
+        new_centroids = xr.DataArray(da.from_array(res), dims=['y', 'x', 'xyz'])
+
+        # Back-conversion to geographic CRS
+        lonlatalt = da.map_blocks(self._do_transform, geocent, latlong,
+                                  new_centroids[:, :, 0].data,  # x
+                                  new_centroids[:, :, 1].data,  # y
+                                  new_centroids[:, :, 2].data,  # z
+                                  new_axis=[2],
+                                  chunks=new_centroids.data.chunks)
+
+        # Back-conversion array as input format
+        lons, _ = _convert_2D_array(lonlatalt[:, :, 0], to=src_lons_format, dims=['y', 'x'])
+        lats, _ = _convert_2D_array(lonlatalt[:, :, 1], to=src_lats_format, dims=['y', 'x'])
+
+        # Add additional info if the source array is a DataArray
+        if isinstance(self.lats, xr.DataArray) and isinstance(self.lons, xr.DataArray):
+            lats.attrs = self.lats.attrs.copy()
+            lons.attrs = self.lons.attrs.copy()
+            try:
+                resolution = lons.attrs['resolution'] / ((x + y) / 2)
+                lons.attrs['resolution'] = resolution
+                lats.attrs['resolution'] = resolution
+            except KeyError:
+                pass
+
+        # Return the downsampled swath definition
+        return SwathDefinition(lons, lats)
+
+    def extend(self, x=0, y=0):
+        """Extend the swath definition along x (along-track) and y (across-track) dimensions.
+
+        By default, it does not extend on any direction.
+        To extend of n pixel  on both sides of the across-track direction, call swath_def.extend(x=0, y=2).
+        """
+        import xarray as xr
+
+        # Check input validity
+        x = int(x)
+        y = int(y)
+        if x < 0 or y < 0:
+            raise ValueError('x and y arguments must be positive integers.')
```

Comment:
Please make the error message more informative, like `SwathDefinition.extend needs positive numbers of pixels.`

Reply:
Done